resolver/resolver_test.go

```diff
@@ -481,3 +297,179 @@ func TestLaunchZK(t *testing.T) {
+}
+
+func answers(m *dns.Msg) []dns.RR { return m.Answer }
+func nss(m *dns.Msg) []dns.RR     { return m.Ns }
+func extras(m *dns.Msg) []dns.RR  { return m.Extra }
+
+func nth(i int, get func(*dns.Msg) []dns.RR, asserts ...func(dns.RR) error) func(*dns.Msg) []error {
+	return func(m *dns.Msg) []error {
+		rrs := get(m)
+		if i >= len(rrs) {
+			return []error{fmt.Errorf("RR not found in position %d", i)}
+		}
+		var errs []error
+		for _, assert := range asserts {
+			if err := assert(rrs[i]); err != nil {
+				errs = append(errs, fmt.Errorf("RR #%d: %v", i, err))
+			}
+		}
+		return errs
+	}
+}
+
+func a(hdr dns.RR_Header, ip net.IP) func(dns.RR) error {
+	return equal("A RR", &dns.A{
+		Hdr: hdr,
+		A:   ip.To4(),
+	})
+}
+
+func srv(hdr dns.RR_Header, target string, port, priority, weight uint16) func(dns.RR) error {
+	return equal("SRV RR", &dns.SRV{
+		Hdr:      hdr,
+		Target:   target,
+		Port:     port,
+		Priority: priority,
+		Weight:   weight,
+	})
+}
+
+func ns(hdr dns.RR_Header, s string) func(dns.RR) error {
+	return equal("NS", &dns.NS{
+		Hdr: hdr,
+		Ns:  s,
+	})
+}
+
+func soa(hdr dns.RR_Header, ns, mbox string, minttl uint32) func(dns.RR) error {
+	return equal("SOA", &dns.SOA{
+		Hdr:    hdr,
+		Ns:     ns,
+		Mbox:   mbox,
+		Minttl: minttl,
+	})
+}
+
+func equal(desc string, want dns.RR) func(dns.RR) error {
+	return func(rr dns.RR) error {
+		if got := rr; !reflect.DeepEqual(got, want) {
+			return fmt.Errorf("%s: %v", desc, pretty.Compare(got, want))
+		}
+		return nil
+	}
+}
+
+// responseRecorder implements the dns.ResponseWriter interface. It's used in
+// tests only.
+type responseRecorder struct {
+	localAddr  net.IPAddr
+	remoteAddr net.IPAddr
+	msg        *dns.Msg
+}
+
+func (r responseRecorder) LocalAddr() net.Addr  { return &r.localAddr }
+func (r responseRecorder) RemoteAddr() net.Addr { return &r.remoteAddr }
+func (r *responseRecorder) WriteMsg(m *dns.Msg) error {
+	r.msg = m
+	return nil
+}
+
+func (r *responseRecorder) Write([]byte) (int, error) { return 0, nil }
+func (r *responseRecorder) Close() error              { return nil }
+func (r responseRecorder) TsigStatus() error          { return nil }
+func (r responseRecorder) TsigTimersOnly(bool)        {}
+func (r *responseRecorder) Hijack()                   {}
+
+func fakeDNS(t *testing.T) *Resolver {
+	res := New("", records.Config{
+		Masters:    []string{"144.76.157.37:5050"},
+		TTL:        60,
+		Port:       53,
+		Domain:     "mesos",
+		Resolvers:  records.GetLocalDNS(),
+		Listener:   "127.0.0.1",
+		SOARname:   "root.ns1.mesos.",
+		SOAMname:   "ns1.mesos.",
+		HTTPPort:   8123,
+		ExternalOn: true,
+	})
+	res.rng.Seed(0) // for deterministic tests
+
+	b, err := ioutil.ReadFile("../factories/fake.json")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var sj records.StateJSON
+	err = json.Unmarshal(b, &sj)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	masters := []string{"144.76.157.37:5050"}
+	spec := labels.RFC952
+	res.rs = &records.RecordGenerator{}
```

Comment:
Cool, thanks. This function was pretty much left unchanged, but I'll address these comments.